addons/mail/models/res_users.py

```diff
@@ -117,23 +117,19 @@ def _unsubscribe_from_channels(self):
 
     @api.model
     def systray_get_activities(self):
-        activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
-        activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
-        for activity in activities:
-            record = self.env[activity.res_model].browse(activity.res_id)
-            activities_by_record_by_model_name[activity.res_model][record] += activity
-        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_model_name.keys()})
+        activities = self.env["mail.activity"].read_group([("user_id", "=", self.env.uid)], ["res_ids:array_agg(res_id)", "res_model_id"], ["res_model_id"])
         user_activities = {}
-        for model_name, activities_by_record in activities_by_record_by_model_name.items():
-            domain = [("id", "in", list({r.id for r in activities_by_record.keys()}))]
-            allowed_records = self.env[model_name].search(domain)
+        for group in activities:
+            model = self.env["ir.model"].browse(group["res_model_id"][0])
+            domain = [("id", "in", group["res_ids"])]
+            allowed_records = self.env[model.model].search(domain)
             if not allowed_records:
```

Comment:
Just wondering, would it be possible to put "res_ids" > 0 condition inside the read group or not?

Reply:
I don't get why you would put that, if the user has no res_ids for a model, it has no records with res_model_id in the table, so it is already the case?